compiler/rustc_errors/src/lib.rs

```diff
@@ -968,12 +969,12 @@ impl HandlerInner {
     }
 
     fn delay_good_path_bug(&mut self, msg: &str) {
-        let mut diagnostic = Diagnostic::new(Level::Bug, msg);
+        let diagnostic = Diagnostic::new(Level::Bug, msg);
         if self.flags.report_delayed_bugs {
             self.emit_diagnostic(&diagnostic);
         }
-        diagnostic.note(&format!("delayed at {}", std::backtrace::Backtrace::force_capture()));
-        self.delayed_good_path_bugs.push(diagnostic);
+        let backtrace = std::backtrace::Backtrace::force_capture();
```

Comment:
hm, I'm wondering if we should get the chain of functions getting to this point annotated with #[track_caller] and entirely remove the backtrace capture in favor of that information - the capture itself is still pretty expensive, so avoiding it may be worthwhile.

Reply:
I'm just doing peephole optimizations here, but my vague understanding of this logic it that it records possible causes for failures and only emits them later when compilation actually fails. If compilation succeeds then they're discarded. Which means track-caller doesn't help, since there's no panic at the point that emits the message. The panic (or other kind of error) that causes these traces to be recalled happens at some later point.

Reply:
I don't understand. My guess is the "delayed at" message need not retain the full backtrace; we only need the direct caller in practice for 99% of cases, such that the compiler developer aiming to fix a delayed bug can identify the location and (for example) replace it with a non-good-path bug, or directly fix it.

If this is true, then we should be able to use the https://doc.rust-lang.org/nightly/std/panic/struct.Location.html#method.caller method to capture a "shallow" 1-point backtrace, using track_caller to propagate it up as needed, which should be much cheaper than capturing a full backtrace.

Reply:
Ah, I wasn't aware of that method and thought caller information would only be used in panics.  `Location` contains file and line-number.
Is that determined at compile-time or would that need runtime parsing of debuginfo similar to backtraces?

If it's determined at compile time then it might work but I don't know what would be the correct depth for the `#[track_caller]` annotations. There only seems to be one callsite for this.

https://github.com/rust-lang/rust/blob/bacf770f2983a52f31e3537db5f0fe1ef2eaa874/compiler/rustc_middle/src/ty/print/pretty.rs#L2277-L2298

Reply:
It was introduced in #73996

Reply:
All compile-time, no runtime parsing currently. The track_caller annotation likely needs to go all the way up to that method (not including the one in print::pretty), likely here in error and on the wrapper(s?) in session.

Reply:
from the [commit](https://github.com/rust-lang/rust/pull/73996/commits/7b2deb562822112cc30d23958e7459564e2c6ef9) introducing it:

> Our current user of this method is deeply nested, being a side effect
from `Display` formatting on lots of rustc types. So taking only the
caller to the error message is not enough - we should collect the
traceback instead.

Reply:
Hmm. I am trying to understand the *purpose* of the Backtrace here. If it is to debug, perhaps it should be gated behind debug_assertions, or at least behind RUST_BACKTRACE or a similar environment variable? It's not clear to me why we would *require* a backtrace collection always; that seems unnecessary.

It may be worth getting a measurement from a simple full removal of the backtrace collection here: if it's not a big win, then there's no point in worrying about anything further here. If it is, we can continue exploring.

cc @petrochenkov perhaps you remember context from that PR?

Reply:
Commenting out `diagnostic.note(&format!("delayed at {}", std::backtrace::Backtrace::force_capture()));` on master and rerunning the UI tests with debuginfo=2 takes 33.96s compared to the lazy version's 34.07s.

So most of the speedup comes from skipping symbolication, taking the backtrace itself is fairly cheap.

Reply:
Sounds good then. Thanks for doing this investigation! Presuming perf.rlo comes back looking good, then I'm happy to merge.